src/automata/include/automata/ta_product.hpp

```diff
@@ -22,64 +22,134 @@
-	}
-	for (const auto &clock : ta2.get_clocks()) {
-		res.add_clock(clock);
+	if (automata.empty()) {
+		throw std::invalid_argument("Cannot compute product of zero automata");
 	}
-	for (const auto &[location, transition] : ta1.get_transitions()) {
-		for (const auto &l2 : ta2.get_locations()) {
-			res.add_transition(Transition{ProductLocation{{location, l2}},
-			                              transition.symbol_,
-			                              ProductLocation{{transition.target_, l2}},
-			                              transition.clock_constraints_,
-			                              transition.clock_resets_});
+	for (auto it1 = std::begin(automata); it1 != std::prev(std::end(automata)); ++it1) {
+		for (auto it2 = std::next(it1); it2 != std::end(automata); ++it2) {
+			std::set<std::string> common_clocks;
+			std::set_intersection(begin(it1->get_clocks()),
+			                      end(it1->get_clocks()),
+			                      begin(it2->get_clocks()),
+			                      end(it2->get_clocks()),
+			                      std::inserter(common_clocks, end(common_clocks)));
+
+			if (!common_clocks.empty()) {
+				throw std::invalid_argument(
+				  fmt::format("Cannot construct product automaton for two automata with non-disjoint "
+				              "clocks, common clocks: {}",
+				              fmt::join(common_clocks, ", ")));
+			}
 		}
 	}
-	for (const auto &[location, transition] : ta2.get_transitions()) {
-		for (const auto &l1 : ta1.get_locations()) {
-			res.add_transition(Transition{ProductLocation{{l1, location}},
-			                              transition.symbol_,
-			                              ProductLocation{{l1, transition.target_}},
-			                              transition.clock_constraints_,
-			                              transition.clock_resets_});
-		}
+	using ProductLocation = Location<std::vector<LocationT>>;
+	std::set<ActionT>         actions;
+	std::set<ProductLocation> product_locations;
+	// Initialize the product locations with the locations from the first TA, by creating a set of
+	// Location<std::vector<LocationT>> from the set of Location<LocationT>.
+	std::transform(std::begin(std::begin(automata)->get_locations()),
+	               std::end(std::begin(automata)->get_locations()),
+	               std::inserter(product_locations, end(product_locations)),
+	               [](const auto &location) { return ProductLocation{{location.get()}}; });
+	// Add the locations from the remaining TAs.
+	std::for_each(std::next(std::begin(automata)),
+	              std::end(automata),
+	              [&product_locations](const auto &ta) {
+		              std::set<ProductLocation> augmented_product_locations;
+		              for (const auto &product_location : product_locations) {
+			              for (const auto &ta_location : ta.get_locations()) {
+				              auto new_product_location = product_location;
+				              new_product_location->push_back(ta_location);
```

Comment:
What exactly does push_back do here? Or, more precisely, ta::Location is a strong type, what does push_back do in this case?

Reply:
Wait, got it. Location just adds properties, the type is std::vector<LocationT>